packages/path_provider/path_provider_platform_interface/lib/src/method_channel_path_provider.dart

```diff
@@ -0,0 +1,90 @@
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'enums.dart';
+
+import 'package:flutter/services.dart';
+import 'package:meta/meta.dart';
+import 'package:path_provider_platform_interface/path_provider_platform_interface.dart';
+import 'package:platform/platform.dart';
+
+/// An implementation of [PathProviderPlatform] that uses method channels.
+class MethodChannelPathProvider extends PathProviderPlatform {
+  /// The method channel used to interact with the native platform.
+  @visibleForTesting
+  MethodChannel methodChannel =
+      MethodChannel('plugins.flutter.io/path_provider');
+
+  // Ideally, this property shouldn't exist, and each platform should
+  // just implement the supported methods. Once all the platforms are
+  // federated, this property should be removed.
+  Platform _platform = const LocalPlatform();
+
+  /// This API is only exposed for the unit tests. It should not be used by
+  /// any code outside of the plugin itself.
+  @visibleForTesting
+  void setMockPathProviderPlatform(Platform platform) {
+    _platform = platform;
+  }
+
+  Future<String> getTemporaryPath() async {
+    return methodChannel.invokeMethod<String>('getTemporaryDirectory');
+  }
+
+  Future<String> getApplicationSupportPath() async {
+    return methodChannel.invokeMethod<String>('getApplicationSupportDirectory');
+  }
+
+  Future<String> getLibraryPath() async {
+    if (!_platform.isIOS && !_platform.isMacOS) {
+      throw UnsupportedError('Functionality only available on iOS/macOS');
+    }
+    return methodChannel.invokeMethod<String>('getLibraryDirectory');
+  }
+
+  Future<String> getApplicationDocumentsPath() async {
+    return methodChannel
+        .invokeMethod<String>('getApplicationDocumentsDirectory');
+  }
+
+  Future<String> getExternalStoragePath() async {
+    if (!_platform.isAndroid) {
+      throw UnsupportedError('Functionality only available on Android');
+    }
+    return methodChannel.invokeMethod<String>('getStorageDirectory');
+  }
+
+  Future<List<String>> getExternalCachePaths() async {
+    if (!_platform.isAndroid) {
+      throw UnsupportedError('Functionality only available on Android');
+    }
+    final List<String> paths = await methodChannel
+        .invokeListMethod<String>('getExternalCacheDirectories');
+
+    return paths.map((String path) => path).toList();
+  }
+
+  Future<List<String>> getExternalStoragePaths({
+    AndroidStorageDirectory type,
+  }) async {
+    if (!_platform.isAndroid) {
+      throw UnsupportedError('Functionality only available on Android');
+    }
+    final List<String> paths = await methodChannel.invokeListMethod<String>(
+      'getExternalStorageDirectories',
+      <String, dynamic>{'type': type?.index},
+    );
+
+    return paths.map((String path) => path).toList();
+  }
+
+  Future<String> getDownloadsPath() async {
+    if (!_platform.isMacOS) {
+      throw UnsupportedError('Functionality only available on macOS');
+    }
```

Comment:
Same comment I did in your previous platform interface migration. IMO this should be thrown by the Java side (I don't know what to do with the 'getDownloadsDirectory' call, and not in the MethodChannel implementation).

If the Fuchsia version is able to fulfill this, this'd be throwing an exception before it even reaches native code.

Reply:
Not sure I follow.  I'm keeping it here to make sure it's as similar to the current implementation as possible. I think you are right about the fact that the native side would just return `MissingImplementation`. I think if we want that exception as opposed to the `UnsupportedError` that might be a bigger discussion to be consistent with the rest of the plugins.

What do you mean by the fuchsia version example? In theory, nothing other than macOS should reach the native code. Maybe that's why the platform checks were added? To save some calls if we know they are going to fail? Also, maybe for testing :)

Reply:
Let's agree to disagree :P

Reply:
Oh I don't have a strong posture on this at all :) It's mostly a "Want to change as little as possible from the way it already was". I think @amirh or @cyanglaz  might have a better opinion on this.